crates/sui-benchmark/src/bin/shared.rs

```diff
@@ -82,272 +91,308 @@ async fn main() {
     // We will create as many counters as the number of ops
     let num_shared_counters = max_in_flight_ops;
 
+    // create server runtime
+    let server_runtime = Builder::new_multi_thread()
+        .thread_stack_size(32 * 1024 * 1024)
+        .worker_threads(opts.num_server_threads)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    // create client runtime
+    let client_runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
+        .worker_threads(opts.num_client_threads)
+        .build()
+        .unwrap();
+
+    // Make the client runtime wait until we are done creating genesis objects
+    let barrier = Arc::new(Barrier::new(2));
+    let cloned_barrier = barrier.clone();
+
     // Create enough gas objects to cover for creating and incrementing counters
     let mut gas = vec![];
     let mut counters_gas = generate_gas_objects(num_shared_counters);
     let publish_module_gas = generate_gas_object();
     gas.append(&mut counters_gas);
     gas.push(publish_module_gas.clone());
+    let gas_cloned = gas.clone();
 
     // Setup the network
-    let configs = test_and_configure_authority_configs(opts.committee_size);
-    let _ = spawn_test_authorities(gas.clone(), &configs).await;
-    let clients = test_authority_aggregator(&configs);
-    let quorum_driver_handler = QuorumDriverHandler::new(clients.clone());
+    let configs = Arc::new(test_and_configure_authority_configs(opts.committee_size));
+    let cloned_configs = configs.clone();
 
-    // publish package
-    write("Publishing basics package".to_string());
-    let package_ref = publish_counter_package(publish_module_gas, configs.validator_set()).await;
-    gas.pop();
+    // spawn a thread to spin up sui nodes on the multi-threaded server runtime
+    let _ = std::thread::spawn(move || {
```

Comment:
As you know, that the thread used here doesn't impose a hierarchy on your tokio executor, i.e. as you mention on top of the PR, this does not impose a resource consumption limit on either the children of this thread, or those not among its descendents. If that was what you wanted to do, I'd expect you'd reach for a ThreadPool-ed executor.

Reply:
The spawning of single thread is only needed to spawn the server runtime (otherwise tokio wouldn't let me spawn client and the server runtime both from the main thread, it complains about nested runtimes). The idea is to spin up server and its children tokio tasks all on the server runtime. And then do the same for the client on the client runtime.